mlflow/gateway/providers/mlflow.py

```diff
@@ -0,0 +1,207 @@
+            raise HTTPException(
+                status_code=502,
+                detail=f"The response is missing the required key: {MLFLOW_SERVING_RESPONSE_KEY}.",
+            )
+        return response[MLFLOW_SERVING_RESPONSE_KEY]
+
+    @staticmethod
+    def _process_payload(payload, key):
+        payload = jsonable_encoder(payload, exclude_none=True)
+
+        input_data = payload.pop(key, None)
+        request_payload = {"inputs": input_data if isinstance(input_data, list) else [input_data]}
+
+        if payload:
+            request_payload["params"] = payload
+
+        return request_payload
+
+    def _process_completions_response_for_mlflow_serving(self, response):
+        inference_data = self._extract_mlflow_response_key(response)
+
+        if isinstance(inference_data, dict):
+            inference_data = inference_data.get("candidates", inference_data)
+
+        is_list_of_strings = isinstance(inference_data, list) and all(
+            isinstance(item, str) for item in inference_data
+        )
+
+        if not is_list_of_strings:
+            raise HTTPException(
+                status_code=502,
+                detail=(
+                    "The response structure from the MLflow serving endpoint is not "
+                    "compatible with the completions route type. The value components must be "
+                    "a list of strings."
+                ),
+            )
+
+        return [{"text": entry, "metadata": {}} for entry in inference_data]
+
+    async def completions(self, payload: completions.RequestPayload) -> completions.ResponsePayload:
+        # Example request to MLflow REST API server for completions:
+        # {
+        #     "inputs": ["hi", "hello", "bye"],
+        #     "params": {
+        #         "temperature": 0.5,
+        #         "top_k": 3,
+        #     }
+        # }
+
+        resp = await send_request(
+            headers=self.headers,
+            base_url=self.mlflow_config.model_server_url,
+            path="invocations",
+            payload=self._process_payload(payload, "prompt"),
+        )
+
+        # Example response:
+        # {"predictions": ["hello", "hi", "goodbye"]}
+
+        return completions.ResponsePayload(
+            **{
+                "candidates": self._process_completions_response_for_mlflow_serving(resp),
+                "metadata": {
+                    "model": self.config.model.name,
+                    "route_type": self.config.route_type,
+                },
+            }
+        )
+
+    def _process_chat_response_for_mlflow_serving(self, response):
+        inference_data = self._extract_mlflow_response_key(response)
+
+        if isinstance(inference_data, dict):
+            inference_data = inference_data.get("candidates", inference_data)
+
+        is_list_of_strings = isinstance(inference_data, list) and all(
+            isinstance(item, str) for item in inference_data
+        )
```

Comment:
Would be great to have pydantic models for this kind of stuff to simplify validation, but not urgent for this PR

Reply:
agreed on using pydantic validation. made the change and it simplified the code (loads of free validation for what it can validate) and testing is more robust.